### PR TITLE
Add final where appropriate

### PR DESCRIPTION
**Pull request by practicalswift**

### CODE REVIEW

**Comment by theuni** [comment]
concept ACK, but please prefer final where possible

**Comment by practicalswift** [comment]
@theuni Done :-)

**Comment by MarcoFalke** [comment]
same binaries on 9d847d2474a229f7a10d6e5fffd433aba7ce85cc, fyi

File: src/test/coins_tests.cpp [hunk 1]
```diff
@@ -28,7 +28,7 @@ class CCoinsViewTest : public CCoinsView
     std::map<uint256, CCoins> map_;
 
 public:
-    bool GetCoins(const uint256& txid, CCoins& coins) const
+    bool GetCoins(const uint256& txid, CCoins& coins) const final
```
Comment (sipa) [inline src/test/coins_tests.cpp]: Isn't it easier to mark the entire class CCoinsViewTest as final?

**Comment by sipa** [comment]
I verified that with GCC 6.2 at -O2, a final on either a method or a class can avoid a vtable lookup, so concept ACK for sure.

I don't think we have a need for specifying final on a per-method bases, though - all examples you are applying this to can use class finals.

Other classes: CCoinsViewCache (coins.h), CCoinsViewDB (txdb.h), CCoinsViewMempool (txmempool.h), CWallet (wallet/wallet.h).

**Comment by morcos** [comment]
@sipa how does that mesh with @marcofalke's observation of same binaries?   Different gcc?

I'd prefer if we decide to do this that we add some documentation that it's just for performance reasons and that if people change the code so there are non obvious requirements for a derived class that they should document that. It seems to me final is sometimes used to indicate don't footgun yourself by trying to derive from this.

**Comment by theuni** [comment]
@morcos Agreed, and I think that a comment at the location of the "final" in that case would suffice. For ex:
```
class Derived : Base
{
    foo() final // unsafe for derivation. Do not change to 'override'.
};

class Fickle final // deriving from Fickle would be unsafe.
{
};
```
There's no way for me to override foo() or Final without removing those warnings. If nothing else, it would be painfully obvious in review, and a change would have to be paired with a justification.

**Comment by sipa** [comment]
@morcos I only tested with isolated example, where class B derives from class A, and an overridden method is called on a variable of type B that comes from another module. As the compiler can't know that B isn't a reference to an object of a class C that derives from B, it has to emit the vtable lookup. But in many cases, analysis can show that this B variable points to an object of type B exactly (especially if it was constructed within the same module). If adding `final` does not change the binary, I guess that for all cases where it was introduced, the compiler was already able to infer enough. Note that there are more classes were `final` could apply.

**Comment by laanwj** [comment]
Let's not just randomly pepper `final` behind method signatures because we can.

> There's no way for me to override foo() or Fickle without removing those warnings. If nothing else, it would be painfully obvious in review, and a change would have to be paired with a justification.

This sounds like a better use of `final`. Document *why* on a per-case basis.
And prefer doing it per class instead of for every single method.

**Pull request closed without merging**